diff --git a/configure/RELEASE.local b/configure/RELEASE.local
index 52eae13..bbf873c 100644
--- a/configure/RELEASE.local
+++ b/configure/RELEASE.local
@@ -20,8 +20,8 @@
 # could match a directory name.
 # ==========================================================
 ASYN_MODULE_VERSION=R4.30-1.1.1
-#CALC_MODULE_VERSION=R3.3-0.1.1
-#SSCAN_MODULE_VERSION=R2.9-1.2.0
+CALC_MODULE_VERSION=R3.3-0.1.1
+SSCAN_MODULE_VERSION=R2.9-1.2.0
 
 # ==========================================================
 # Define module paths using pattern
@@ -30,8 +30,8 @@ ASYN_MODULE_VERSION=R4.30-1.1.1
 # FOO = /Full/Path/To/Development/Version 
 # ==========================================================
 ASYN=$(EPICS_MODULES)/asyn/$(ASYN_MODULE_VERSION)
-#CALC=$(EPICS_MODULES)/calc/$(CALC_MODULE_VERSION)
-#SSCAN=$(EPICS_MODULES)/sscan/$(SSCAN_MODULE_VERSION)
+CALC=$(EPICS_MODULES)/calc/$(CALC_MODULE_VERSION)
+SSCAN=$(EPICS_MODULES)/sscan/$(SSCAN_MODULE_VERSION)
 
 # Set EPICS_BASE last so it appears last in the DB, DBD, INCLUDE, and LIB search paths
 EPICS_BASE              = $(EPICS_SITE_TOP)/base/$(BASE_MODULE_VERSION)
diff --git a/streamApp/Makefile b/streamApp/Makefile
index de1f038..6cba610 100644
--- a/streamApp/Makefile
+++ b/streamApp/Makefile
@@ -13,7 +13,7 @@ include $(TOP)/configure/CONFIG
 
 #HOST_OPT = NO
 
-PROD = streamApp
+#PROD = streamApp
 DBD  = streamApp.dbd
 
 streamApp_DBD += base.dbd
@@ -42,7 +42,7 @@ ifneq ($(words $(CALC) $(SYNAPPS)), 0)
 # With synApps scalcout record
 streamApp_DBD += calcSupport.dbd
 streamApp_DBD += streamSynApps.dbd
-#PROD_LIBS += calc
+PROD_LIBS += calc
 # older calc versions require sscan
 PROD_LIBS += sscan
 PROD_LIBS_DEFAULT += streamSynApps
diff --git a/streamApp/test.cmd b/streamApp/test.cmd
old mode 100644
new mode 100755
